assemblyscript.html.markdown

```diff
@@ -0,0 +1,228 @@
+let f3 = (i: i32): i32 => i * i;
+
+// Classes - members are public by default
+class Point {
+  // Properties
+  x: f64;
+
+  // Constructor - the public/private keywords in this context will generate
+  // the boiler plate code for the property and the initialization in the
+  // constructor.
+  // In this example, "y" will be defined just like "x" is, but with less code
+  // Default values are also supported
+
+  constructor(x: f64, public y: f64 = 0) {
+    this.x = x;
+  }
+
+  // Functions
+  dist(): f64 { return Math.sqrt(this.x * this.x + this.y * this.y); }
+
+  // Static members
+  static origin: Point = new Point(0, 0);
+}
+
+// Classes can be explicitly marked as implementing an interface.
+// Any missing properties will then cause an error at compile-time.
+class PointPerson implements Person {
+    name: string
+    move(): void {}
+}
+
+let p1 = new Point(10, 20);
+let p2 = new Point(25); //y will be 0
+
+// Inheritance
+class Point3D extends Point {
+  constructor(x: f64, y: f64, public z: f64 = 0) {
+    super(x, y); // Explicit call to the super class constructor is mandatory
+  }
+
+  // Overwrite
+  dist(): f64 {
+    let d = super.dist();
+    return Math.sqrt(d * d + this.z * this.z);
+  }
+}
+
+// Modules, "." can be used as separator for sub modules
+module Geometry {
+  class Square {
+    constructor(public sideLength: f64 = 0) {
+    }
+    area(): void {
+      return Math.pow(this.sideLength, 2);
+    }
+  }
+}
+
+let s1 = new Geometry.Square(5);
+
+// Generics
+// AssemblyScript compiles generics to one concrete method or function per set 
+// of unique contextual type arguments, also known as [monomorphisation]. 
+// Implications are that a module only includes and exports concrete functions 
+// for sets of type arguments actually used and that concrete functions can be 
+// shortcutted with [static type checks] at compile time, which turned out to 
+// be quite useful.
+// Classes
+class Tuple<T1, T2> {
+  constructor(public item1: T1, public item2: T2) {
+  }
+}
+
+// Interfaces
+interface Pair<T> {
+  item1: T;
+  item2: T;
+}
+
+// And functions
+function pairToTuple <T>(p: Pair<T>): Tuple<T, T> {
+  return new Tuple(p.item1, p.item2);
+};
+
+let tuple = pairToTuple<string>({ item1: "hello", item2: "world" });
+
+// Including references to a definition file:
+/// <reference path="jquery.d.ts" />
+
+// Template Strings (strings that use backticks)
+// String Interpolation with Template Strings
+let name = 'Tyrone';
+let greeting = `Hi ${name}, how are you?`
+// Multiline Strings with Template Strings
+let multiline = `This is an example
+of a multiline string`;
+
+let numbers: Array<i8> = [0, 1, 2, 3, 4];
+let moreNumbers: Array<i8> = numbers;
+moreNumbers[5] = 5; // Error, elements are read-only
+moreNumbers.push(5); // Error, no push method (because it mutates array)
+moreNumbers.length = 3; // Error, length is read-only
+numbers = moreNumbers; // Error, mutating methods are missing
```

Comment:
Maybe just show how an array can be used:
```suggestion
numbers[5] = 5; // Set element at an index
numbers.push(5); // Add an element to the end of the array
```

Reply:
It is already good as it is but thanks for suggesting.

Reply:
What's currently there is just not true.

Reply:
Yes, that's why it is commented it produces errors. The same in Typescript

Reply:
No it doesn't produce errors, since it's valid